activemodel/lib/active_model/error.rb

```diff
@@ -0,0 +1,197 @@
+        return msg
+      end
+
+      i18n_options = {
+        default: msg || defaults,
+        model: @base.model_name.human,
+        attribute: humanized_attribute,
+        value: value,
+        object: @base
+      }.merge!(options)
+
+      I18n.translate(key, i18n_options)
+    end
+
+    # Returns a full message for a given attribute.
+    #
+    #   person.errors.first.full_message # => "Name is invalid"
+    #
+    #   The `"%{attribute} %{message}"` error format can be overridden with either
+    #
+    # * <tt>activemodel.errors.models.person/contacts/addresses.attributes.street.format</tt>
+    # * <tt>activemodel.errors.models.person/contacts/addresses.format</tt>
+    # * <tt>activemodel.errors.models.person.attributes.name.format</tt>
+    # * <tt>activemodel.errors.models.person.format</tt>
+    # * <tt>errors.format</tt>
+    def full_message
+      message = self.message
+
+      return message if @attribute == :base
+
+      if ActiveModel::Errors.i18n_full_message && @base.class.respond_to?(:i18n_scope)
+        parts = attribute_without_index.split(".")
+        attribute_name = parts.pop
+        namespace = parts.join("/") unless parts.empty?
+        attributes_scope = "#{@base.class.i18n_scope}.errors.models"
+
+        if namespace
+          defaults = @base.class.lookup_ancestors.map do |klass|
+            [
+              :"#{attributes_scope}.#{klass.model_name.i18n_key}/#{namespace}.attributes.#{attribute_name}.format",
+              :"#{attributes_scope}.#{klass.model_name.i18n_key}/#{namespace}.format",
+            ]
+          end
+        else
+          defaults = @base.class.lookup_ancestors.map do |klass|
+            [
+              :"#{attributes_scope}.#{klass.model_name.i18n_key}.attributes.#{attribute_name}.format",
+              :"#{attributes_scope}.#{klass.model_name.i18n_key}.format",
+            ]
+          end
+        end
+
+        defaults.flatten!
+      else
+        defaults = []
+      end
+
+      defaults << :"errors.format"
+      defaults << "%{attribute} %{message}"
+
+      I18n.t(defaults.shift,
+        default: defaults,
+        attribute: humanized_attribute,
+        message: message)
+    end
+
+    # See if error matches provided +attribute+, +type+ and +options+.
+    def match?(attribute, type = nil, **options)
+      if @attribute != attribute || (type && @type != type)
+        return false
+      end
+
+      options.each do |key, value|
+        if @options[key] != value
+          return false
+        end
+      end
+
+      true
+    end
+
+    def strict_match?(attribute, type, **options)
+      return false unless match?(attribute, type, **options)
+
+      full_message == Error.new(@base, attribute, type, **options).full_message
+    end
+
+    def ==(other)
+      attributes_for_hash == other.attributes_for_hash
+    end
+    alias eql? ==
```

Comment:
This probably needs to check the type, or a `respond_to?`.  I think code like this might raise an exception:

```ruby
if error_object == nil
  ...
end
```

I know it seems like a contrived example, but nobody would expect `==` to raise ever.

Reply:
Ah, I would have never thought of that!